src/oauth.rs

```diff
@@ -0,0 +1,105 @@
+use crate::settings;
+use crate::token::{verify_jwt_token_from_rsa, Claims};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct Response {
+    pub email: String,
+    pub claims: Claims,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Key {
+    #[serde(skip)]
+    kty: String,
+
+    pub n: String, // We only need n and e
+    pub e: String,
+
+    #[serde(skip)]
+    qi: String,
+
+    #[serde(skip)]
+    kid: String,
+
+    #[serde(skip)]
+    alg: String,
+
+    #[serde(skip)]
+    r#use: String,
+
+    #[serde(skip)]
+    #[serde(rename = "fxa-createdAt")]
+    fxa_created_at: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct JWK {
+    pub keys: Vec<Key>,
+}
+
+fn scope_matches(scope: &String) -> bool {
+    // TODO: complete function as available at url:
+    // https://github.com/mozilla/PyFxA/blob/53a9b649dd1225a641be95ea2ab3e241533ef562/fxa/_utils.py#L68
+    true
+}
```

Comment:
I am having little difficulty implementing this function. If you look at the  [python code](https://github.com/mozilla-services/tokenserver-rs/pull/17#issuecomment-702286197) scope have to be list so I was thinking I can make this a tuple or Vec[String] but I am having problem with the matching part. There are two arguments required and provided and I am not able to give provide as an argument.

Reply:
Sure. Looking at the Python code, the Required Scopes are passed along [as an optional argument](https://github.com/mozilla/PyFxA/blob/7c341e75b6bc47c3b44a36a6975208db6fbf03b7/fxa/oauth.py#L257) for `verify`, so you should add them as an argument in rust. 

For this I'd change the argument list to be a `provided:Vec<String>, required: Option<Vec<String>>`, which would be the simplest to deal with the borrow checker. 

Inside of `verify`, we don't have to type check what we're getting (yay! rust!) but we do need to check to see if the provided scopes match up against what we were told are required. You can do that using something like

```rust
if let Some(required_scopes) = required {
   for provided_scope in provided {
       for required_scope in required_scopes {
            if match_single_scope(provided_scope, required_scope) {
                   return true
            }
      }
}
return false
```
* Please check that code. I wrote it just now, on the fly, so I can't guarantee it will work right. ;) 

The `match_single_scope` is a bit more complicated. Each scope is either a string of URLs like:
`https://identity.mozilla.com/apps/oldsync#read` or short words like `profile:write`. Since you're running low on time, let's not deal with all the rules and conditions for now. For URLs, just have it return `false` with a `//TODO: finish this` comment (see below if you want to create a better, more useful comment). For short words, there are parts, separated by a `:`. The provided parts must match what's in the required list, so for example:

`foo` would match required `foo:bar`
`foo:bar` would match required `foo:bar`
`foo:gorp` would NOT match required `foo:bar`
and
`bar` would NOT match required `foo:bar` 

Feel free to add appropriate "//TODO: " comments for the parts you don't feel you can finish off. Try to make them useful for the next person so that they can understand what needs to be done. (Try to imagine yourself coming back to this code in 6 months and needing to work on it again. What would be the things you'd need to read to pick up where you left off?) 

Comments in code like this are invaluable, particularly when you need to potentially hand off code to someone else. This happens because of a lot of reasons. A contract might be ending early, or you could be reassigned or promoted, or what-have-you. Likewise, you might be handed code from someone else and now it's your task to get it working. Putting good comments in, particularly around things you know need work or might be tricky makes the next person's life a lot easier, and sometimes, that future person is you. No reason not to be nice to yourself.

Reply:
Thanks for such a wonderful insight John. This was very helpful. I have implemented the `scope_matches` function.  However, I am unable to fix why `test_jwks` is failing barring that all tests are running fine. Can you look into that?